src/objective/rank_obj.cu

```diff
@@ -0,0 +1,637 @@
+  return CountNumLabelsImpl(false, labels, n, v);
+}
+
+class SortedLabelList {
+ private:
+  // Labels sorted within the group
+  dh::caching_device_vector<float> dsorted_labels_;
+
+  // Original position of the labels before they are sorted descendingly within its groups
+  dh::caching_device_vector<uint32_t> doriginal_pos_;
+
+  // Segments within the original list that delineates the different groups
+  dh::caching_device_vector<uint32_t> group_segments_;
+
+  // Need this on the device as it is used in the kernels
+  dh::caching_device_vector<uint32_t> dgroups_;       // Group information on device
+
+  int device_id_{-1};                                 // GPU device ID
+  const LambdaRankParam &param_;                      // Objective configuration
+
+ public:
+  SortedLabelList(int dev_id,
+                  const LambdaRankParam &param)
+    : device_id_(dev_id),
+      param_(param) {}
+
+  void InitWithTrainingInfo(const std::vector<uint32_t> &groups) {
+    int num_elems = groups.back();
+
+    dsorted_labels_.resize(num_elems);
+
+    doriginal_pos_.resize(num_elems);
+    thrust::sequence(doriginal_pos_.begin(), doriginal_pos_.end());
+
+    group_segments_.resize(num_elems);
+
+    dgroups_ = groups;
+
+    // Launch a kernel that populates the segment information for the different groups
+    uint32_t *gsegs = group_segments_.data().get();
+    const unsigned *dgroups = dgroups_.data().get();
+    int ngroups = dgroups_.size();
+    dh::LaunchN(device_id_, num_elems, nullptr, [=] __device__(unsigned idx){
+      // Find the group first
+      int group_idx = dh::UpperBound(dgroups, ngroups, idx);
+      gsegs[idx] = group_idx - 1;
+    });
+  }
+
+  // Sort the groups by labels. We would like to avoid using predicates to perform the sort,
+  // as thrust resorts to using a merge sort as opposed to a much much faster radix sort
+  // when comparators are used. Hence, the following algorithm is used. This is done so that
+  // we can grab the appropriate prediction values from the original list later, after the
+  // labels are sorted.
+  //
+  // Here is the internal representation:
+  // dgroups_:          [ 0, 3, 5, 8, 10 ]
+  // group_segments_:   0 0 0 | 1 1 | 2 2 2 | 3 3
+  // doriginal_pos_:    0 1 2 | 3 4 | 5 6 7 | 8 9
+  // dsorted_labels_:   1 0 1 | 2 1 | 1 3 3 | 4 4 (from original labels)
+  //
+  // Sort the labels first and make a note of the original positions in doriginal_pos_
+  // based on the sort
+  // dsorted_labels_:   4 4 3 3 2 1 1 1 1 0
+  // doriginal_pos_:    8 9 6 7 3 0 2 4 5 1
+  // NOTE: This consumes space, but is much faster than some of the other approaches - sorting
+  //       in kernel, sorting using predicates etc.
+  void Sort(const HostDeviceVector<bst_float> &dlabels) {
+    dsorted_labels_.assign(dh::tcbegin(dlabels), dh::tcend(dlabels));
+    thrust::stable_sort_by_key(dsorted_labels_.begin(), dsorted_labels_.end(),
```

Comment:
You should pass a temporary allocator to thrust sort like this: https://github.com/thrust/thrust/blob/c7eee649fabd91896cbd10adc8f69ea77bf83b44/examples/cuda/custom_temporary_allocation.cu#L173

I would expect this to considerably improve performance.

Reply:
done.

jfyi - comparing this with the previous numbers i published earlier, the gradient computation time for the small mslr dataset comes down from 0.5s to 0.3s and for the large one from 1.2s to .9s.